### PR TITLE
Merge membership Monitor and ServiceResolver to membership.Resolver

### PR DESCRIPTION
**Pull request by mantas-sidlauskas**
**What changed?**
Membership monitor and service resolver interfaces merged to a single `membership.Resolver` interface.

* `Resolver` is using hashring to get ownership for a shard. 
* Hashring is using members provided by ringpop. 

**Why?**
This is part of refactoring needed to introduce selectable membership providers.

**How did you test it?**
Unit test/local test

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: service/frontend/adminHandler.go [hunk 1]
```diff
@@ -597,24 +597,22 @@ func (adh *adminHandlerImpl) DescribeCluster(
 
 		var rings []*types.RingInfo
 		for _, role := range service.List {
-			resolver, err := monitor.GetResolver(role)
-			if err != nil {
-				return nil, adh.error(err, scope)
```
Comment (vytautas-karpavicius) [inline service/frontend/adminHandler.go:602]: Seems like we no longer return this error. Is this intended?

Comment (mantas-sidlauskas) [reply service/frontend/adminHandler.go:602]: thanks, fixed

File: client/matching/peerResolver.go [hunk 2]
```diff
@@ -20,28 +20,31 @@
...
+	"github.com/uber/cadence/common/membership"
+	"github.com/uber/cadence/common/service"
+)
 
 // PeerResolver is used to resolve matching peers.
 // Those are deployed instances of Cadence matching services that participate in the cluster ring.
 // The resulting peer is simply an address of form ip:port where RPC calls can be routed to.
 type PeerResolver struct {
-	membership    membership.ServiceResolver
+	membership    membership.Resolver
```
Comment (vytautas-karpavicius) [inline client/matching/peerResolver.go:32]: nit: this was renamed to `resolver` in `client/history/peerResolver.go`, maybe rename here as well for consistency?

Comment (mantas-sidlauskas) [reply client/matching/peerResolver.go:32]: yep, can do

File: service/history/shard/controller.go [hunk 3]
```diff
@@ -163,7 +165,7 @@ func (c *controller) Start() {
 	c.shutdownWG.Add(1)
 	go c.shardManagementPump()
 
-	err := c.GetHistoryServiceResolver().AddListener(shardControllerMembershipUpdateListenerName, c.membershipUpdateCh)
+	err := c.GetMembershipResolver().Subscribe(service.History, shardControllerMembershipUpdateListenerName, c.membershipUpdateCh)
 	if err != nil {
 		c.logger.Error("Error adding listener", tag.Error(err))
```
Comment (vytautas-karpavicius) [inline service/history/shard/controller.go:170]: Please update this error message, as this is not listener anymore, but a subscriber.

File: service/history/shard/controller.go [hunk 4]
```diff
@@ -178,7 +180,7 @@ func (c *controller) Stop() {
 
 	c.PrepareToStop()
 
-	if err := c.GetHistoryServiceResolver().RemoveListener(shardControllerMembershipUpdateListenerName); err != nil {
+	if err := c.GetMembershipResolver().Unsubscribe(service.History, shardControllerMembershipUpdateListenerName); err != nil {
 		c.logger.Error("Error removing membership update listener", tag.Error(err), tag.OperationFailed)
```
Comment (vytautas-karpavicius) [inline service/history/shard/controller.go:184]: Update this error as well, listener -> subscriber.

File: service/matching/matchingEngine.go [hunk 5]
```diff
@@ -109,7 +111,7 @@ func NewEngine(taskManager persistence.TaskManager,
 	logger log.Logger,
 	metricsClient metrics.Client,
 	domainCache cache.DomainCache,
-	resolver membership.ServiceResolver,
+	monitor membership.Resolver,
```
Comment (vytautas-karpavicius) [inline service/matching/matchingEngine.go:114]: This var should be `resolver`

File: service/worker/replicator/domain_replication_processor.go [hunk 6]
```diff
@@ -55,7 +76,7 @@ func newDomainReplicationProcessor(
 	metricsClient metrics.Client,
 	taskExecutor domain.ReplicationTaskExecutor,
 	hostInfo *membership.HostInfo,
-	serviceResolver membership.ServiceResolver,
+	membership membership.Resolver,
```
Comment (vytautas-karpavicius) [inline service/worker/replicator/domain_replication_processor.go:79]: nit: rename to `resolver` for consistency with other places

File: common/membership/hashring.go [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+		event.HostsAdded = append(event.HostsAdded, NewHostInfo(addr, r.getLabelsMap()))
+	}
+	for _, addr := range rpEvent.ServersRemoved {
+		event.HostsRemoved = append(event.HostsRemoved, NewHostInfo(addr, r.getLabelsMap()))
+	}
+	for _, addr := range rpEvent.ServersUpdated {
+		event.HostsUpdated = append(event.HostsUpdated, NewHostInfo(addr, r.getLabelsMap()))
+	}
+
+	// Notify listeners
```
Comment (vytautas-karpavicius) [inline common/membership/hashring.go:269]: listeners -> subscribers

**Review by vytautas-karpavicius — APPROVED** [review]

**Pull request merged**